LIBCLOUD-269 : Multipart upload for amazon S3

mahendra opened the pull request:
This patch adds support for streaming data upload using Amazon's multipart upload support as listed in [S3 docs](http://docs.amazonwebservices.com/AmazonS3/latest/dev/UsingRESTAPImpUpload.html)

As per current behaviour, the `upload_object_via_stream()` API will collect the entire object in memory, and then upload it to S3. This can turn problematic with large files (think HD videos around 4GB). This will be a huge hit in performance and memory of the python application.

With this patch, the API `upload_object_via_stream()` will use the S3 multipart upload feature to upload data in 5MB chunks, thus reducing the overall memory impact of the application.
## Design of this feature:
- The `S3StorageDriver()` is not used just for Amazon S3. It is sub-classed for being used with other S3 compliant cloud storage providers like Google Storage.
- The Amazon S3 multipart upload is not (or may not be) supported by other storage providers (who will prefer the chunked upload mechanism)

We can solve this problem in two ways:
- Create a new subclass of `S3StorageDriver` (say `AmazonS3StorageDriver`), which implements this new multipart upload mechanism. Other storage providers will subclass `S3StorageDriver`. This is a more cleaner approach.
- Introduce an attribute `supports_s3_multipart_upload` and based on it's value, control the callback function passed to `_put_object()` API. This makes the code look a bit hacky, but this approach is better for supporting such features in the future. We don't have to keep making sub-classes for each feature.

In the current patch, I have implemented the latter approach, though I prefer the former. After discussions with the community and knowing their preferences, we can select a final approach.
## Design notes:
- Implementation has three steps
  1. `POST` request to `/container/object_name?uploads`. This returns an XML with a unique `uploadId`. This is handled as part of `_put_object()`. Doing it via `_put_object()` ensures that all S3 related parameters are set correctly.
  2. Upload each chunk via `PUT` to `/container/object_name?partNumber=X&uploadId=***` - This is done via the callback that is passed to `_put_object()` named `_upload_multipart()`
  3. `POST` an XML containing part-numbers and etag headers returned for each part to `/container/object_name?uploadId=***`, implemented via `_commit_multipart()`
  4. In case of any failures in steps (2) or (3), the upload is deleted from S3 through a `DELETE` request to `/container/object_name?uploadId=****`, implemented via `_abort_multipart()`
- The chunk size for upload was set as 5MB - This is the minimum allowed size as per Amazon S3 docs.
## Other changes:
- Did some PEP8 cleanup on s3.py
- `s3.get_container()` would iterate through the list of containers for finding the requested entry. This can be simplified by making a `HEAD` request. The only downside is that `created_time` is not available for the container. Let me know if this approach is OK or if I must revert it.
- Introduced the following APIs for the `S3StorageDriver()`, to make some functionality easier.
  - `get_container_cdn_url()`
  - `get_object_cdn_url()`
- In `libcloud.common.base.Connection`, the `request()` method is used as the basis for all HTTP requests made by libcloud. This method had a limitation, which became apparent in S3 multipart upload implementation. For initializing an upload, the API invoked was  `/container/object_name?uploads`.  The `uploads` parameter had to be passed as-is, without any values. If we made use of `params` argument in `request()` method, it would have come up as `uploads=***`. To prevent this, the `action` was set to `/container/object_name?uploads` and slight modifications were made to how parameters were appended. This also forced a change in `BaseMockHttpObject._get_method_name()`
## Bug fixes in test framework
- While working on the test cases, I noticed a small issue. Not sure if it was a bug or as per design. `MockRawResponse._get_response_if_not_availale()` would return two different values on subsequent invocations.

```
     if not self._response:
         ...
         return self  <----- this was inconsistent.
     return self._response
```

  While adding test cases for the Amazon S3 functionality, I noticed that instead of getting back `MockResponse`, I was getting `MockRawResponse` instance (which did not have methods like `read()` or `parse_body()`). So, I fixed this issue. Because of this other test cases started failing and they were subsequently fixed. Not sure if this has to be fixed or if it was done on purpose. If someone can throw some light on it, I can work on it further. As of now, all test cases pass.
- In test_s3.py, the driver was being set everywhere to `S3StorageDriver`. This same test case is used for `GoogleStorageDriver`, where the driver turns up as `S3StorageDriver` instead of `GoogleStorageDriver`. This was fixed by changing code to `driver=self.driver_type`

mahendra committed:
LIBCLOUD-261 : Iterator based API for container listing

mahendra committed:
LIBCLOUD-265 : In local storage, remove empty parent folders

mahendra committed:
Merge branch 'trunk' into hash

Conflicts:
	libcloud/storage/drivers/cloudfiles.py
	libcloud/storage/drivers/nimbus.py

mahendra committed:
Merge branch 'trunk' into hash

mahendra committed:
Merge branch 'trunk' into hash

mahendra committed:
LIBCLOUD-269 : Multipart upload for amazon S3

Kami commented on libcloud/common/base.py:
```diff
@@ -574,7 +574,10 @@ def request(self,
 
         if params:
-            url = '?'.join((action, urlencode(params)))
+            if '?' in action:
```
Would be nice to document this behavior in the docstring.

mahendra replied:
Done.

Kami replied:
I don't see any new commits yet. Looks like you still need to push the changes :)

mahendra replied:
Sorry. I was addressing comments one-by-one. Did a push after I answered all the comments. You should be able to see one now.

Kami commented on libcloud/storage/drivers/s3.py:
```diff
@@ -234,6 +241,42 @@ def get_object(self, container_name, object_name):
+        if check:
+            self.get_container(container.name)
+
+        return '/%s' % (container.name)
```
This method shouldn't really be called `get_container_cdn_url` because it just returns an internal path to the object.

`get_container_cdn_url` is supposed to return a full CDN url of a container. In AWS CDN is handled by a separate service (CloudFront) so adding CDN support is more involved.

mahendra replied:
Agreed. Changed it to _get_container_path()

Kami commented on libcloud/storage/drivers/s3.py:
```diff
@@ -234,6 +241,42 @@ def get_object(self, container_name, object_name):
+        object_path = '%s/%s' % (container_url, object_name_cleaned)
+        return object_path
+
+    def get_object_cdn_url(self, obj):
```
Same here.

mahendra replied:
Changed it to _get_object_path()

Kami commented on libcloud/storage/drivers/s3.py:
```diff
@@ -337,6 +373,133 @@ def upload_object(self, file_path, container, object_name, extra=None,
+
+            # Commit the chunk info and complete the upload
+            etag = self._commit_multipart(object_path, upload_id, chunks)
+        except Exception:
```
Hm, can we catch more specific exception here?

In any case we should at least include the original exception message in the custom exception we re-throw.

mahendra replied:
I thought about it. In Amazon multipart upload, the uploadId keeps lingering if the upload is not aborted explicitly. 

Using boto, we can observe all the pending multipart uploads.

```
connection = boto.s3.S3Connection(...)
bucket = connection.get_bucket(container_name)
uploads = bucket.get_all_multipart_uploads()
```

So, Ideally we should abort the multipart upload in case of any exception. Else, we will leave lingering uploads pending in S3. That is why, I decided to catch a generic exception.

Will change the code to ensure that the specific exception is thrown back

Kami replied:
Looks good now :)

Kami replied:
Another thing - can we also add an extension method for listing lingering multipart uploads (e.g. `ex_list_multipart_uploads)?

Kami replied:
Besides this and a couple of other minor things I have mentioned, the patch looks great.

Please update JIRA ticket with the latest patch when you are done with addressing those things and I will proceed with testing and merging it into trunk :)

mahendra replied:
Yeah sure. Will add few more methods

```
ex_list_multipart_uploads
ex_abort_multipart_upload
ex_cleanup_all_multipart_uploads
```

Also, need to check how to expose these in the CLI commands. Have some ideas. Will work on a future ticket for this.
Once I implement this, will upload patch to JIRA

mahendra replied:
Taken care of all your comments. Thanks for the detailed review.

Implemented the following extensions

```
ex_iterate_multipart_uploads()
ex_cleanup_all_multipart_uploads()
```

Kami commented on libcloud/storage/drivers/s3.py:
```diff
@@ -337,6 +373,133 @@ def upload_object(self, file_path, container, object_name, extra=None,
+        else:
+            encoding = None
+
+        data = tostring(root, encoding=encoding)
```
Hm, utility function which wraps original string and uses correct encoded based on the Python version should probably live in `libcloud.utils.py3`.

mahendra replied:
Hmmm, could not find any... Or do you want me to create one there?

Kami replied:
Sorry, I wasn't clear in my previous comment. One doesn't exist yet so it would be great if you can create it there.

mahendra replied:
OK. So, you want it to be like this in utils.py3

```
if sys.version_info >= (3, 0):
    encoding = 'unicode'
else:
    encoding = None
```

and it being used as 
`ET.tostring(data, encoding=encoding)`

OR

```
if sys.version_info >= (3,0):
    ...
    def tostring(data):
        return ET.tostring(data)

else:
    tostring = ET.tostring
```

(ideally the second one should be in utils.xml.py
OR, we can use a combination of both. (using 'xml_encoding' in py3) and tostring() in xml.py

Kami replied:
Number two or a combination of both.

mahendra replied:
Actually, this is turning a bit tricky because of import conflict among `utils.xml` and `xml.extree.ElementTree`

mahendra replied:
Is it ok, if we rename the xml.py to xmlutils.py or something?

Kami replied:
I'm a bit confused, can you please clarify what's conflicting?

Can't you just use an alias when importing it?

mahendra replied:
In the code, if I use

`from xml.etree import ....`

it clashes with

```
libcloud.utils.xml
```

Kami commented on libcloud/storage/drivers/s3.py:
```diff
@@ -337,6 +373,133 @@ def upload_object(self, file_path, container, object_name, extra=None,
+
+    def _upload_from_iterator(self, iterator, object_path, upload_id,
+                              calculate_hash=True):
+        """Uploads data from an interator in fixed sized chunks to S3"""
```
Please add docstrings for all the arguments (http://libcloud.apache.org/docstring-conventions.html).

mahendra replied:
Done

mahendra committed:
LIBCLOUD-269 : Address review comments

Kami commented on libcloud/test/storage/test_s3.py:
```diff
@@ -617,6 +736,34 @@ def test_upload_object_via_stream(self):
+                                                       object_name=object_name,
+                                                       iterator=iterator,
+                                                       extra=extra)
+        except Exception:
```
Need to use sys "hack" otherwise it won't work with all the supported Python versions.

mahendra committed:
LIBCLOUD-269 : Implement driver specific extensions

Driver specific extensions for handling multipart uploads
* ex_iterate_multipart_uploads()
* ex_cleanup_all_multipart_uploads()

Also address review comments

mahendra committed:
LIBCLOUD-269 : Test cases for the multipart extensions

mahendra commented:
Fixed by https://issues.apache.org/jira/browse/LIBCLOUD-269
